reworked Trait class and moved all trait init code to archetypes module

feend78 opened the pull request:
This pull request addresses issues #33  and #34.
- Flattened out some parts of the world directory; moved traits.py and archetypes.py into world
- Reworked and enhanced Trait class significantly.
- Removed nearly all trait management code from the Character typeclass (mod_stat, etc).
- Updated Character typeclass `at_object_create` hook, and added the `traits` property as described in the traits module.
- Moved individual archetype classes into archetypes module
- Updated individual archetype class definitions to reflect new traits configuration
- Created and updated module functions that will support chargen process: `load_archetype`, `apply_archetype`, `validate_primary_traits`, and `calculate_secondary_traits`
- Changed functions in `rulebook` module to more accurately reflect OA dice rolls.
- Created `sheet` and `traits` commands to view full character sheet or traits by group.
- Docs and tests for everything I touched. 

As we still don't have a chargen process built, there is some setup required to see the output of the commands. From inside Ainneve, you can initialize your character with:

```
@typeclass me = typeclasses.characters.Character
@py from world.archetypes import apply_archetype; apply_archetype(me, 'warrior')
@py me.traits.STR.base += 3; me.traits.PER.base += 1
@py me.traits.DEX.base += 1; me.traits.VIT.base += 3
@py from world.archetypes import calculate_secondary_traits; calculate_secondary_traits(me.traits)
sheet
```

feend78 committed:
reworked Trait class and moved all trait init code to archetypes module
created 'sheet' and 'traits' commands to view character traits
docs and tests for everything

feend78 committed:
modified TraitFactory to throw an exception if user tries to assign a value to a trait key
added a test for this condition and consolidated some tests

ergodicbreak commented on world/archetypes.py:
```diff
@@ -0,0 +1,279 @@
+        the same character with two different `name` parameters to create
+        a dual archetype. May also be called with reset=True to remove
+        any existing archetype and initialize the character with only the
+        named one.
```
cool, I like this

ergodicbreak commented on world/traits.py:
```diff
@@ -0,0 +1,701 @@
+    All traits have a read-only `actual` property that will report the trait's
+    actual value. In addition to normal attribute notation, `Trait` objects
+    support the unary `+` operator as a shortcut to access this property.
+
```
while this is cool, is there a compelling use case for it that justifies the added code/complexity?

feend78 replied:
The use cases were driven both by OA rules and after consideration of common categories of trait behaviors I've seen in many games. There are examples of all three trait types in the base trait configuration in `__init__` for the `Archetype` class.

In working on this, I was informed by Ainneve's use cases, but was also trying to make the `Trait` class something general purpose and robust enough that it could be easily reused in others' games. That's also part of why I went so over the top with the documentation.

ergodicbreak replied:
Apologies, the comment is meant only for the `+` operator

feend78 replied:
As we discussed in IRC the other day, the implementation of the `+` operator and other comparison/arithmetic magic methods are not really crucial to the functioning of the Trait class, but I put them there because I personally prefer the briefer syntax that they allow over the more explicit and verbose use of the `actual` property. However, I can appreciate the desire for the explicit form in the context of Ainneve as an example for beginner programmers. Do you want me to take it all out?

ergodicbreak replied:
Yes, I think the magic + operator isn't a good beginner example.

I think there's no harm in leaving it in the code (it has an explanatory comment already -- a nice Easter egg if you will), but I would prefer that it isn't highlighted in the documentation.

ergodicbreak commented on world/traits.py:
```diff
@@ -0,0 +1,701 @@
+            if self.max is not None and value >= self.max:
+                return self.max
+        return value
+
```
really nice work on this module, thank you!

feend78 committed:
removed `+` operator documentation
replaced `+` shortcuts with `.actual`
updated tests
minor tweak to charsheet layout
added a README for the logs directory, as the Travis CI report shows errors starting Evennia after checkout because server/logs doesn't exist

feend78 commented:
I removed the `+` operator documentation but left the code there. 

I also added a server/logs/README.md file, because I noticed the Travis CI tests were not running because the logs directory was not present. Now, it looks like the tests are failing because of recent changes to sessions in core Evennia that are throwing errors in the `evennia.utils.test_resources.EvenniaTest` class used in the test modules.

feend78 committed:
modified hook calls that were expecting sessid instead of session

feend78 commented:
So I'm curious about the next steps. I assume @whitenoiseoss should review it, but then is it ready to merge?

ergodicbreak commented:
Thanks for resolving the testing issue; I think this is ready, I'll leave it open today for @whitenoiseoss but if he doesn't have time to look at it today I can merge tonight. 

It looks like your other PRs are finished as well?

feend78 commented:
I think #47 is ready too, but I kinda want to pull the starter equipment one back and resubmit it later. Now that there's a basis for working with Traits, I went ahead and started working on weapon/armor effects and abilities, and I'd like to include them before moving ahead with that one.

feend78 committed:
added another useful module function to archetypes - get_remaining_allocation

feend78 commented:
Since this is still open, I want to include this one other idea I had that I think will be needed in character generation, and that the presumably EvMenu implementation shouldn't have to be calculating. It's fairly simple, and in line with keeping all Trait-related logic in archetypes.

whitenoiseoss commented:
I have merged this over giving it a fairly thorough look-through. For the OA rulebook, moving the traits off of character and onto archetype is a good call and will make things easier. The expansion to Traits is great.
